domain/cloudimagemetadata/state/state.go

```diff
@@ -0,0 +1,381 @@
+	// requiring an housekeeping goroutine. Since a metadata data doesn't live long in the db and
+	// it shouldn't be a lot of custom metadata, it shouldn't slow too much the insertion.
+	// Expired metadata are ignored when retrieving metadata, so it doesn't have any impact to keep
+	// some of them longer than their time to live.
+	if err := s.tryCleanUpExpiredMetadata(ctx); err != nil {
+		return errors.Trace(err)
+	}
+	if err := s.ensuresArchitecturesLoaded(ctx); err != nil {
+		return errors.Trace(err)
+	}
+
+	db, err := s.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Prepare inputs
+	createdAt := time.Now()
+	values := make([]inputMetadata, 0, len(metadatas))
+	seen := make(map[string]bool)
+	for _, m := range metadatas {
+		err := validateMetadata(m)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		// Handle duplicated inputs
+		key := buildKey(m)
+		if seen[key] {
+			return cloudmetadataerrors.MultipleInsertNonUnique
+		}
+		seen[key] = true
+
+		// Convert architecture name to a db id
+		archId, ok := s.architectureIDsByName[m.Arch]
+		if !ok {
+			s.logger.Warningf("skipping metadata for unknown architecture %q", m.Arch)
+			continue
+		}
+
+		// Update creation time if not defined.
+		var noTime time.Time
+		if m.CreationTime != noTime {
+			createdAt = m.CreationTime
+		}
+
+		values = append(values, inputMetadata{
+			Key:             key,
+			CreatedAt:       createdAt,
+			Source:          m.Source,
+			Stream:          m.Stream,
+			Region:          m.Region,
+			Version:         m.Version,
+			VirtType:        m.VirtType,
+			ArchitectureID:  archId,
+			RootStorageType: m.RootStorageType,
+			RootStorageSize: m.RootStorageSize,
+			Priority:        m.Priority,
+			ImageID:         m.ImageId,
+		})
+	}
+
+	// Prepare statement for checking if an image metadata already exists in DB, to handle update
+	existMetadataStmt, err := sqlair.Prepare(`
+SELECT &metadataKey.unique_key FROM cloud_image_metadata WHERE unique_key = $metadataKey.unique_key `, metadataKey{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Prepare statement to update an existing image in DB (it only update the imageID
+	updateMetadataStmt, err := sqlair.Prepare(`
+UPDATE cloud_image_metadata
+SET image_id = $metadataImageID.image_id
+WHERE unique_key = $metadataKey.unique_key`, metadataImageID{}, metadataKey{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Prepare statement to insert a batch of non existing image in db
+	insertMetadataStmt, err := sqlair.Prepare(`
+INSERT INTO cloud_image_metadata (*)
+VALUES ($inputMetadata.*);`, inputMetadata{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return errors.Trace(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Define for each values if it should be updated or inserted
+		updated := make([]inputMetadata, 0, len(values))
+		inserted := make([]inputMetadata, 0, len(values))
+
+		for _, v := range values {
```

Comment:
Can we replace this for loop with a `WHERE ... IN` sql clause?

Reply:
No because the updated images are updated independantly. However, we could rely on a `ON CONFLICT DO UPDATE`, but some project prevent using this kind of clause, so it didn't come to me at first intent.

However I see that it is used in juju code at several places, so I will rewrite it accordingly, I have to rewrite this method anyway to get ride of the `buildkey` function anyway

Reply:
OK cool, I see that now you do a batch insert, that's all I wanted actually. Thanks